compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -0,0 +1,608 @@
+    applyPermutationToVector(strides, invPerm);
+    SmallVector<int64_t> baseShape = applyPermutation(srcShape, invPerm);
+
+    // Check if the resulting offsets, sizes, and strides correspond to a
+    // contiguous slice and can thus be mappable to a `flow.tensor.update` op.
+    // This should always be worth doing because this can remove a dispatch for
+    // the slice, and the transpose is on the slice rather than the full tensor.
+    if (!IREE::Flow::isOffsetSizeAndStrideMappableToFlow(offsets, sizes,
+                                                         strides, baseShape)) {
+      return failure();
+    }
+
+    ArrayRef<int64_t> staticSizes = extractOp.getStaticSizes();
+    ArrayRef<int64_t> sliceShape = extractOp.getResultType().getShape();
+    llvm::SmallDenseSet<unsigned> rankReducingMask =
+        *mlir::computeRankReductionMask(staticSizes, sliceShape);
+
+    int64_t dim = 0;
+    llvm::SmallDenseMap<int64_t, int64_t> rankReducedMap;
+    for (int64_t i = 0, e = perm.size(); i < e; ++i) {
+      if (!rankReducingMask.contains(i)) {
+        rankReducedMap[i] = dim++;
+      }
+    }
+
+    // Compute the new permutation by dropping all rank-reduced dimensions.
+    SmallVector<int64_t> rankReducedPerm;
+    for (int64_t i : perm) {
+      if (!rankReducingMask.contains(i)) {
+        rankReducedPerm.push_back(rankReducedMap[i]);
+      }
+    }
+
+    auto rankReducedInvPerm = invertPermutationVector(rankReducedPerm);
+
+    RankedTensorType sliceType = getPermutedTensorType(
+        cast<RankedTensorType>(extractOp.getType()), rankReducedInvPerm);
+    Value slice = rewriter.create<tensor::ExtractSliceOp>(
+        extractOp.getLoc(), sliceType, transposeOp.getInput(), offsets, sizes,
+        strides);
+    // Transpose back to the original slice.
+    if (!isIdentityPermutation(rankReducedPerm)) {
+      slice = createTranspose(rewriter, slice, rankReducedPerm);
+    }
+    rewriter.replaceOp(extractOp, slice);
+    return success();
+  }
+};
+
+// Sinks a transpose through a tensor.expand_shape.
+class SinkTransposeThroughExpandShape
+    : public OpRewritePattern<tensor::ExpandShapeOp> {
+public:
+  using OpRewritePattern<tensor::ExpandShapeOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::ExpandShapeOp expandOp,
+                                PatternRewriter &rewriter) const override {
+    if (!IREE::Flow::isNonNullAndOutsideDispatch(expandOp)) {
+      return failure();
+    }
+    Value source = expandOp.getSrc();
+    auto transposeOp = source.getDefiningOp<linalg::TransposeOp>();
+    // Do not propagate through reshapes if the transpose has multiple users, as
+    // this could end up duplicating the transposes. We should only propagate
+    // through reshape when it is free to do so.
+    if (!transposeOp || !transposeOp->hasOneUse()) {
+      return failure();
+    }
+
+    ArrayRef<int64_t> perm = transposeOp.getPermutation();
+
+    auto invPerm = invertPermutationVector(perm);
```

Comment:
nit: maybe we can combine these into one line. The `perm` is only used by`invPerm`.

```suggestion
    auto invPerm = invertPermutationVector(transposeOp.getPermutation());
```